### PR TITLE
239 organization client interaction space editing

### PR DESCRIPTION
**Pull request by mmcdermott**

### CODE REVIEW

File: src/lib/engine/space.js [hunk 1]
```diff
@@ -33,15 +33,25 @@ const user = (space, graph) => {
   return graph.users.find(e => sameIds(e.id, space.user_id))
 }
 
+const organization = (space, graph) => {
+  return graph.organizations.find(e => sameIds(e.id, space.organization_id))
+}
+
+const isMember = (organization_id, user_id, memberships) => (
```
Comment (OAGr) [inline src/lib/engine/space.js]: why is this in ``space.rb``?

File: src/lib/engine/space.js [hunk 2]
```diff
@@ -33,15 +33,25 @@ const user = (space, graph) => {
...
+  !!_.find(memberships, m => (m.organization_id === organization_id && m.user_id === user_id))
```
Comment (OAGr) [inline src/lib/engine/space.js]: instead of ``_.find`` you can do ``_.some``

**Comment by mmcdermott** [comment]
So, there's a bug with this, where it basically doesn't try to fetch all the spaces users to display them with the spaces: 

user-missing

My sense is this is lower priority than getting space creation to work, and we shouldn't focus on this for now. I can also just disable showing users at all for this page until we get it to work properly.

File: src/lib/engine/space.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import * as _graph from './graph';
 import * as _metric from './metric';
 import * as _guesstimate from './guesstimate';
+import * as UserOrganizationMemberships from 'gEngine/userOrganizationMemberships'
```
Comment (OAGr) [inline src/lib/engine/space.js]: this should be imported like the stuff above.

File: src/lib/engine/space.js [hunk 4]
```diff
@@ -33,15 +34,21 @@ const user = (space, graph) => {
...
+
 export function toDgraph(spaceId, graph){
   let dGraph = _graph.denormalize(subset(graph, spaceId))
   const space = get(graph.spaces, spaceId)
   const spaceUser = user(space, graph)
+  const userOrganizationMemberships = graph.userOrganizationMemberships
+  const meId = _.get(graph, 'me.id')
   dGraph.user = spaceUser
-  dGraph.ownedByMe = sameIds(_.get(spaceUser, 'id'), _.get(graph, 'me.id'))
+  dGraph.ownedByMe = sameIds(space.user_id, meId) || UserOrganizationMemberships.isMember(space.organization_id, meId, userOrganizationMemberships)
```
Comment (OAGr) [inline src/lib/engine/space.js]: I would split this into a separate function, with userOrganizationMemberships, which only seems useful for this line.

File: src/lib/engine/space.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
-import * as _graph from './graph';
-import * as _metric from './metric';
-import * as _guesstimate from './guesstimate';
+import * as _graph from './graph'
+import * as _metric from './metric'
+import * as _guesstimate from './guesstimate'
+import * as _UserOrganizationMemberships from './userOrganizationMemberships'
```
Comment (OAGr) [inline src/lib/engine/space.js]: first character should be lowercase

File: src/modules/spaces/actions.js [hunk 6]
```diff
@@ -57,6 +58,17 @@ export function fetchById(spaceId) {
         const user_id = value.user_id
         const has_user = !!(users.find(e => e.id === user_id))
         if (!has_user) { dispatch(userActions.fetchById(user_id)) }
+
+        // TODO(matthew): Right now, the space has an embedded user and organization record... why are we doing this
```
Comment (OAGr) [inline src/modules/spaces/actions.js]: Looking into it a bit, I'm pretty sure that came from before we had the embedded user, so is no longer needed

File: src/modules/spaces/actions.js [hunk 7]
```diff
@@ -57,6 +58,17 @@ export function fetchById(spaceId) {
...
+        // extra fetching?
+        if (value.organization_id) {
+          const organizations = getState().organizations
+          const organization_id = value.organization_id
+          const has_organization = !!(organizations.find(e => e.id === organization_id))
```
Comment (OAGr) [inline src/modules/spaces/actions.js]: or ``_.some``

File: src/modules/spaces/actions.js (hunk 6 above)
Comment (mmcdermott) [inline src/modules/spaces/actions.js]: Filed #246 to look into this later.

File: src/modules/spaces/actions.js [hunk 8]
```diff
@@ -57,6 +58,17 @@ export function fetchById(spaceId) {
         const user_id = value.user_id
         const has_user = !!(users.find(e => e.id === user_id))
         if (!has_user) { dispatch(userActions.fetchById(user_id)) }
+
+        // TODO(matthew): Right now, the space has an embedded user and organization record... why are we doing this
+        // extra fetching?
+        if (value.organization_id) {
```
Comment (OAGr) [inline src/modules/spaces/actions.js]: I realize that this wasn't done before, but now that this function is so long, I think i would appreciate splitting it up into a helper like ``needs_organization(value.organization_id, getState().organizations)``

File: src/modules/userOrganizationMemberships/actions.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import {actionCreatorsFor} from 'redux-crud'
+import $ from 'jquery' // TODO(matthew): Is this needed at all?
```
Comment (OAGr) [inline src/modules/userOrganizationMemberships/actions.js]: it's not needed unless used in that file

File: src/modules/userOrganizationMemberships/actions.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+import * as displayErrorsActions from 'gModules/displayErrors/actions.js'
+import * as userActions from 'gModules/users/actions.js'
+import {rootUrl} from 'servers/guesstimate-api/constants.js'
+import {captureApiError} from 'lib/errors/index.js'
+import {setupGuesstimateApi} from 'servers/guesstimate-api/constants.js'
+
+let sActions = actionCreatorsFor('userOrganizationMemberships')
+
+function api(state) {
```
Comment (OAGr) [inline src/modules/userOrganizationMemberships/actions.js]: don't worry about this now, but this definitely needs to get extracted sometime

File: src/modules/users/actions.js [hunk 11]
```diff
@@ -84,3 +84,9 @@ export function create(object) {
   }
 }
 
+export function fetchSuccess(users) {
+  return (dispatch) => {
+    const formatted = users.map(d => _.pick(d, ['auth0_id', 'id', 'name', 'picture']))
```
Comment (OAGr) [inline src/modules/users/actions.js]: the ``allowedAttributes`` or whatever should be a new function at this point.  It's repeated above.

File: src/modules/spaces/actions.js [hunk 12]
```diff
@@ -52,11 +63,10 @@ export function fetchById(spaceId) {
       }
       else if (value) {
         dispatch(sActions.fetchSuccess([value]))
-
-        const users = getState().users
-        const user_id = value.user_id
-        const has_user = !!(users.find(e => e.id === user_id))
-        if (!has_user) { dispatch(userActions.fetchById(user_id)) }
+        // TODO(matthew): Right now, the space has an embedded user and organization record... why are we doing this
+        // extra fetching?
```
Comment (OAGr) [inline src/modules/spaces/actions.js]: it may well be unnecessary now, but we may want to merge this code before refactoring that.

**Pull request merged**